### PR TITLE
Make UsbSerial work with multi-interface devices. (e.g. FTDI4232)

### LINKED ISSUE #968: FTDI4232 (4-port USB to RS232) -- only the first port available on Android platform
Chip datasheet:
https://ftdichip.com/products/ft4232hq/ 

XCSoar versions having and not having the problem
-------------------------------------------------
Linux is not affected since it natively identifies the FT4232 ports as ttyUSB[0-3]
Android Version 7.23 is affected

Expected behavior
-----------------
In the device/port dialog all four channels of the FT4232 should show up as separate ports so each can be assigned to an individual driver, so that multiple RS232 devices can be connected.

Actual behavior
---------------
In the device/port dialog only the first channel of the FT4232 is available (labeled "Quad RS232-HS(FTDI)" ).
This port works as expected. But the remaining three ports do not show up and therefore are unavailable.

Steps to reproduce the behavior
-------------------------------
Connect FTDI4232 to USB(-OTG) of an android phone.
I am using this board:
https://www.aliexpress.com/item/1005001928940997.html 

I am of course available for testing new code.

Side note:
Using this: USB-C-(OTG)-PD adapter 
https://de.aliexpress.com/item/1005003887696758.html
the (PD-enabled) phone can get reliably charged while acting as USB-host for the FT4232, so this is a viable setup and will be interesting for many people.
(The phone needs to be capable of USB-C-PD (PowerDeliveryProtocol) though)
Tested on Pixel4a.

Do you have any idea what may have caused this?
-----------------------------------------------
driver implementation of this chip in XCSoar does not account for multiple ports

Do you have an idea how to solve the issue?
-------------------------------------------
update driver code to assign all available ports
might be resolved together with 
https://github.com/XCSoar/XCSoar/issues/966 (not a duplicate though)

### PR DESCRIPTION
**Pull request by henrik1g**
Brief summary of the changes
----------------------------
UsbSerial fully supports multi interface devices. This change makes UsbSerialHelper.java capable providing this functionality to the main program. Some changes were necessary since the code of UsbSerialHelper assumed a one to one relationship of device and port, which is not the case for multiport devices.
Although quite some lines needed changing, the behaviour for single Port devices should not have changed at all.
For multiport devices however entries for each port (=interface) will be available and managed with this change.

Related issues and discussions
------------------------------
https://github.com/XCSoar/XCSoar/issues/968
Closes #968

### CODE REVIEW

**Comment by lordfolken** [comment]
Can you fix the code quality issues, please? (Click on the codacy test to get an overview.)
https://app.codacy.com/gh/XCSoar/XCSoar/pullRequest?prid=10009657&bid=30665931

File: android/src/UsbSerialHelper.java [hunk 1]
```diff
@@ -51,7 +53,31 @@ public class UsbSerialHelper extends BroadcastReceiver {
   private final Context context;
   private final UsbManager usbmanager;
 
-  private HashMap<String, UsbDevice> _AvailableDevices = new HashMap<>();
+  public class UsbDeviceInterface {
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:56]: Should be `static`, and why is it `public`?

File: android/src/UsbSerialHelper.java [hunk 2]
```diff
@@ -51,7 +53,31 @@ public class UsbSerialHelper extends BroadcastReceiver {
...
+    public UsbDevice device;
+    public int iface;
+    public String id;
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:59]: Should be `final`

File: android/src/UsbSerialHelper.java [hunk 3]
```diff
@@ -51,7 +53,31 @@ public class UsbSerialHelper extends BroadcastReceiver {
...
+    public UsbDeviceInterface(UsbDevice dev_,int iface_) {
+      device=dev_;
+      iface=iface_;
+
+      //Alternative id by usb-bus-numbering -- not in use
+      // if (device.getInterfaceCount() > 1) {
+      //   id=device.getDeviceName()+"#"+iface;
+      // } else {
+      //   id=device.getDeviceName();
+      // }
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:69]: What's the point of this comment?

File: android/src/UsbSerialHelper.java [hunk 4]
```diff
@@ -128,27 +154,41 @@ private synchronized void AddAvailable(UsbDevice device) {
 
       if (exists(supported_ids, vid, pid)) {
         Log.v(TAG, "UsbDevice Found : " + device);
-        _AvailableDevices.put(device.getDeviceName(), device);
-        broadcastDetectedDevice(device);
+        for (int iface=0; iface < (device.getInterfaceCount());iface++) {
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:164]: Redundant parantheses, and missing space after `;`

File: android/src/UsbSerialHelper.java [hunk 5]
```diff
@@ -128,27 +154,41 @@ private synchronized void AddAvailable(UsbDevice device) {
...
+          UsbDeviceInterface deviface = new UsbDeviceInterface(device,iface);
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:165]: Missing space after `,`

File: android/src/UsbSerialHelper.java [hunk 6]
```diff
@@ -128,27 +154,41 @@ private synchronized void AddAvailable(UsbDevice device) {
...
-  }
+  private synchronized void RemoveAvailable(UsbDevice removeddevice) {
+    Log.v(TAG,"UsbDevice disconnected : " + removeddevice);
+    // Below line not possible with the current java version
+    //_AvailableInterfaces.entrySet().removeIf(entry -> removeddevice.getDeviceName().equals(entry.getValue().device.getDeviceName()));
+    // Therefore this longer alternative:
+    Iterator<Map.Entry<String,UsbDeviceInterface>> iter = _AvailableInterfaces.entrySet().iterator();
+    while (iter.hasNext()) {
+      Map.Entry<String,UsbDeviceInterface> entry = iter.next();
+      if(removeddevice.getDeviceName().equals(entry.getValue().device.getDeviceName())) {
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:193]: Coding style: missing space after `if`

File: android/src/UsbSerialPort.java [hunk 7]
```diff
@@ -41,10 +41,17 @@
   public UsbSerialPort(UsbDevice device,int baud) {
     _UsbDevice = device;
     _baudRate = baud;
+    _UsbInterface = -1;
+  }
+  public UsbSerialPort(UsbDevice device,int baud,int iface) {
+    _UsbDevice = device;
+    _baudRate = baud;
+    _UsbInterface = iface;
   }
```
Comment (MaxKellermann) [inline android/src/UsbSerialPort.java:50]: Redundant code; both constructors can be merged into one.

**Review by MaxKellermann — CHANGES_REQUESTED** [review]

File: android/src/UsbSerialHelper.java [hunk 8]
```diff
@@ -42,6 +43,7 @@
 
 import com.felhr.usbserial.UsbSerialDevice;
 
+
```
Comment (MaxKellermann) [inline android/src/UsbSerialHelper.java:46]: Why add a blank line here?

**Comment by henrik1g** [comment]
I will fix the issues.
Should I amend my previous commits, or would this break this pull-request process here? What is the preferred way?

File: android/src/UsbSerialPort.java (hunk 7 above)
Comment (henrik1g) [reply android/src/UsbSerialPort.java:50]: I'm not familiar with java. Afai read there are no default parameters, so how would you avoid having two constructors?
Or is this what you suggest?
```
public UsbSerialPort(UsbDevice device,int baud) {
     UsbSerialPort(device, baud, -1);
  }
public UsbSerialPort(UsbDevice device,int baud,int iface) {
    _UsbDevice = device;
    _baudRate = baud;
    _UsbInterface = iface;
  }
```

**Comment by henrik1g** [comment]
I amended the commits in a new branch "UsbSerial_MultiInterfaceV2" let me know if I
- should open a new pull request from the V2 branch
- force push the changed commits on the branch this pull request is for 
- or make an additional "code style" commit on this branch instead

**Comment by MaxKellermann** [comment]
Force-push the amended branch. No new fixup commits, amend the existing commits. Known-bad commits should not be merged.

File: android/src/UsbSerialPort.java (hunk 7 above)
Comment (MaxKellermann) [reply android/src/UsbSerialPort.java:50]: Yes, I'd prefer that over having two ctors with redundant code.

**Comment by henrik1g** [comment]
> Force-push the amended branch. No new fixup commits, amend the existing commits. Known-bad commits should not be merged.

Done

**Comment by henrik1g** [comment]
> Can you fix the code quality issues, please? (Click on the codacy test to get an overview.) https://app.codacy.com/gh/XCSoar/XCSoar/pullRequest?prid=10009657&bid=30665931

All addressed except this one: "Avoid unused private methods such as 'connect(String,int)' "
I would say this is a "false positive" since the methods is definetely used by some other module of xcs, otherwise the code could not work.
If I am missing a point here let me know.

**Comment by MaxKellermann** [comment]
> All addressed except this one: "Avoid unused private methods such as 'connect(String,int)' " I would say this is a "false positive" since the method is definetely used by some other module of xcs, otherwise the code could not work. If I am missing a point here let me know.

Correct, it's used via JNI:
https://github.com/XCSoar/XCSoar/blob/a4ce748a751d453887f297b898819f2656a9800a/src/Android/UsbSerialHelper.cpp#L59-L60
https://github.com/XCSoar/XCSoar/blob/a4ce748a751d453887f297b898819f2656a9800a/src/Android/UsbSerialHelper.cpp#L108-L109

**Comment by MaxKellermann** [comment]
> All addressed

But ... now your first commit contains fixes for old Codacy warnings, which have nothing to do with this PR. Doesn't belong in this commit. Fixing Codacy warnings (if they are correct) is a good idea, but that should be a separate commit.

**Comment by henrik1g** [comment]
> > All addressed
> 
> But ... now your first commit contains fixes for old Codacy warnings, which have nothing to do with this PR. Doesn't belong in this commit. Fixing Codacy warnings (if they are correct) is a good idea, but that should be a separate commit.

I just tried to separate those again... but this proves a little difficult since in some places the changes are on the same line as other changes. 
Can we let that slide for this time? I'll remember it next time...

**Comment by MaxKellermann** [comment]
> I just tried to separate those again... but this proves a little difficult since in some places the changes are on the same line as other changes. Can we let that slide for this time? I'll remember it next time...

Yes, this time. And I suppose you don't know stgit - stgit makes amending patches and authoring a well-designed PR really easy. Without stgit, this is very cumbersome work.

**Comment by henrik1g** [comment]
> > I just tried to separate those again... but this proves a little difficult since in some places the changes are on the same line as other changes. Can we let that slide for this time? I'll remember it next time...
> 
> Yes, this time. And I suppose you don't know stgit - stgit makes amending patches and authoring a well-designed PR really easy. Without stgit, this is very cumbersome work.

In the meantime I got it done as a practice excercise anyway. I'll force push now.
Yes it's cumbersome!

**Comment by hjr** [comment]
Lots of kudos, cannot wait for having those features at hand :) :) :)

**Comment by lyusupov** [comment]
@henrik1g 

This PR have caused a  certain regression of compatibility with other devices in the XCSoar USB 'white list'.

When you add an entry in the Devices list, you ignore it's 'bInterfaceClass' value - which is not good,. An interface of a USB device may not necessary be applicable to perform a data transfer.

Your use case is specific to FTDI4232 only but it needs to be adjusted to meet constraints of other devices in the 'supported' list.

Below is an example of the regression, see the picture below.

S21109-112624

The USB device in the 'white list' has 3 interfaces in total but only the second one is involved in actual 'USB-Serial' data transfers.

I have most of supported USB-Serial adapters readily available here so I did basic research with assistance of '**lsusb** -v' utility in Linux. 

### FT232

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Vendor Specific Class | 255

### FT4232H

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Vendor Specific Class | 255
1 | Vendor Specific Class | 255
2 | Vendor Specific Class | 255
3 | Vendor Specific Class | 255

### CP210x

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Vendor Specific Class | 255

### PL2303

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Vendor Specific Class | 255

### CH34x

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Vendor Specific Class | 255

### CH9102 and other CDC ACM compatible devices

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Communications | 2
1 | CDC Data | 10

### SoftRF Dongle, Academy, Balkan, ES

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Communications | 2
1 | CDC Data | 10

### SoftRF Badge, Lego, Prime Mk3

Interface number | Interface Class | decimal Class Value
:---:|:---:|:---:
0 | Communications | 2
1 | CDC Data | 10
2 | Mass Storage | 8


## Summary

@henrik1g , would you, please, fix the issue by inspecting **bInterfaceClass** value every time you add a device/interface into the list. We should pass the interfaces with class 255 (Vendor Specific Class) and 10 (CDC Data) but ignore any other ones.

Thank you!

**Pull request merged**